### PR TITLE
Improve Autofill search

### LINKED ISSUE #705: [BUG] Oreo autocomplete does not find entry
From #653, cc @msfjarvis 
I just tried the snapshot and the search does not work apparently for the naming scheme path/to/domain.tld.gpg. For example when trying to log into paypal, the search correctly contains paypal.com but it does not find Banking/paypal.com.gpg. Clearing the search and selecting the entry from the list however correctly fills username and password and if i save the association I get it suggested the next time above the option to search

1. Install current snapshot, i used [this link](https://dl.msfjarvis.dev/?dir=APS)
2. Have your passwords organized as multiline files with `path/to/domain.tld.gpg` and the username as second line like `login: <username>`
3. go to a website, I used https://github.com and https://paypal.com in Firefox mobile for android
4. tap into the username or password field to make the oreo autofill ui appear
5. Select "search in store"
6. See search ui with e.g. domain "github.com" pre filled in the search bar
7. See "no results"

**Expected behavior**
A single entry for the site "github.com"

**Workaround**
Clearing the search show all entries as expected, but even clearing and retyping the exact same search string (github.com) leads to the expected one result

**Device information (please complete the following information):**
 - Device: OnePlus 3
 - OS: LineageOS 15.1
 - App version: 1.7-snapshot

### PR DESCRIPTION
**Pull request by FabianHenneke**
## :loudspeaker: Type of change

- [x] Bugfix
- [ ] New feature
- [x] Enhancement
- [ ] Refactoring

## :scroll: Description

These commits aim to make Autofill search more intuitive while preserving its phishing protection. This is done in the following ways:

- In StrictDomain (aka "Phishing-resistant search") mode, we previously only matched a particular path component, governed by the folder structure selected in the settings, against a subdomain of the search term. But users have wildly different directory layouts that we should do our best to support and we should not expect them to tweak settings to get search working. Consequently, StrictDomain now matches any path component against a subdomain of the search term, regardless of the folder structure selected in settings. The origin identifier with respect to this setting is still shown in bold underline.
- In preparation for the next point, and also because it is generally useful, an encrypted username entered for an Autofill generated password is now filled in.
- Support for a folder layout à la "work/Banking/example.org.gpg", which relies on encrypted usernames, is implemented and the ensuing increase in nullability is addressed in the UI and AutofillHelper.
- A switch is added to the Autofill search view which allows switching between Fuzzy and StrictDomain matching for websites.

## :bulb: Motivation and Context

Fixes #705.

## :green_heart: How did you test it?
I tried out the various features that interact with the changed code. @jvanbruegge It would be great if you could also give this a try.

## :pencil: Checklist

- [x] I ran `./gradlew spotlessApply` before submitting the PR
- [x] I reviewed submitted code

## :crystal_ball: Next steps

- [ ] Decide whether the path to the file is useful for distinguishing existing Autofill matches with the new folder layout setting.

### CODE REVIEW

**Comment by jvanbruegge** [comment]
Is there a snapshot build or something similar so I could try it out easily?

**Comment by jvanbruegge** [comment]
Ok, not sure if the artifact really contains the changes because neither do I see the new layout option nor the fuzzy/strict toggle nor does the strict search work

**Comment by FabianHenneke** [comment]
I'm not sure about what the build artifacts contain, but a new snapshot build should be available soon after the PR has been merged. @msfjarvis may be able to provide one even before that happens.

**Comment by msfjarvis** [comment]
> Ok, not sure if the artifact really contains the changes because neither do I see the new layout option nor the fuzzy/strict toggle nor does the strict search work

Something's off there I guess. I'll have to revisit that later. [Here](https://dl.msfjarvis.dev/temp/app-release.apk)'s a signed release build in the mean time.

**Comment by jvanbruegge** [comment]
Thanks, with that build everything works just fine

File: app/src/androidTest/java/com/zeapo/pwdstore/DecryptTest.kt [hunk 1]
```diff
@@ -109,7 +112,19 @@ class DecryptTest {
         // second set the new timer
         activity.settings.edit().putString("general_show_time", "2").commit()
 
-        activity.onBound(null)
+        activity.onBound(object : IOpenPgpService2 {
+            override fun createOutputPipe(p0: Int): ParcelFileDescriptor {
+                TODO("Not yet implemented")
```
[the remark this answers is not in the document]
Comment (FabianHenneke) [reply app/src/androidTest/java/com/zeapo/pwdstore/DecryptTest.kt:117]: Isn't this what we want though? If the test previously worked with a null `IOpenPgpService2` passed in, it certainly shouldn't call any of its methods. What do you think?

[the remark this answers is not in the document]
Comment (msfjarvis) [reply app/src/androidTest/java/com/zeapo/pwdstore/DecryptTest.kt:117]: If the tests are passing then I guess it should be fine. I'm working on enabling instrumentation testing in CI so we'll be able to confirm shortly.

File: app/src/androidTest/java/com/zeapo/pwdstore/StrictDomainRegexTest.kt [hunk 2]
```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright © 2014-2020 The Android Password Store Authors. All Rights Reserved.
+ * SPDX-License-Identifier: GPL-3.0-only
+ */
+package com.zeapo.pwdstore
+
+import kotlin.test.assertFalse
+import kotlin.test.assertNull
+import kotlin.test.assertTrue
+import org.junit.Test as test
```
[the remark this answers is not in the document]
Comment (FabianHenneke) [reply app/src/androidTest/java/com/zeapo/pwdstore/StrictDomainRegexTest.kt:10]: No idea why, but this is actually Jetbrains-official: https://github.com/Kotlin/kotlin-examples/blob/master/gradle/mixed-java-kotlin-hello-world/src/test/kotlin/tests.kt
I changed it anyway.

File: app/src/androidTest/java/com/zeapo/pwdstore/DecryptTest.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (FabianHenneke) [reply app/src/androidTest/java/com/zeapo/pwdstore/DecryptTest.kt:117]: The old ones related to UI and OTPs are no longer passing, but for non-trivial reasons (UI has been changed in the meantime, OTP format is now taken from Steam).

**Comment by FabianHenneke** [comment]
> Just waiting on instrumentation tests to finish [here](https://github.com/android-password-store/Android-Password-Store/actions/runs/79110454) then will submit.

Can we "Rebase and merge" afterwards instead of squashing?

**Comment by msfjarvis** [comment]
> > Just waiting on instrumentation tests to finish [here](https://github.com/android-password-store/Android-Password-Store/actions/runs/79110454) then will submit.
> 
> Can we "Rebase and merge" afterwards instead of squashing?

Sure thing.

**Comment by msfjarvis** [comment]
Welp, instrumentation tests are still failing. I'll submit this and fix those separately.

**Pull request merged**